builder/azure/common/template/template_builder.go

```diff
@@ -316,15 +317,39 @@ func (s *TemplateBuilder) SetOSDiskSizeGB(diskSizeGB int32) error {
 	return nil
 }
 
-func (s *TemplateBuilder) SetDiskEncryptionSetID(diskEncryptionSetID string) error {
+func (s *TemplateBuilder) SetDiskEncryptionSetID(diskEncryptionSetID string, securityType *hashiVMSDK.SecurityTypes) error {
 	resource, err := s.getResourceByType(resourceVirtualMachine)
 	if err != nil {
 		return err
 	}
 
 	profile := resource.Properties.StorageProfile
-	profile.OsDisk.ManagedDisk.DiskEncryptionSet = &DiskEncryptionSetParameters{
-		ID: &diskEncryptionSetID,
+	if securityType != nil && *securityType == hashiVMSDK.SecurityTypesConfidentialVM {
+		profile.OsDisk.ManagedDisk.SecurityProfile = &virtualmachines.VMDiskSecurityProfile{}
+		securityEncryptionType := virtualmachines.SecurityEncryptionTypesDiskWithVMGuestState
```

Comment:
Here and at line 351 you're setting the VM's security ecryption type to DiskWithVMGuestState, whenever I try to build with a CMK while setting a DES, I run into the following error 

```
==> azure-arm.windows-sig: polling after CreateOrUpdate: Code="Conflict" Message="The SecurityProfile type of this image version 'EncryptedWithCmk' must match the SecurityProfile type of the source 'EncryptedVMGuestStateOnlyWithPmk'."
```

When I try to build with EncryptedWithPmk instead I get a similar error
```
==> azure-arm.windows-sig: polling after CreateOrUpdate: Code="Conflict" Message="The SecurityProfile type of this image version 'EncryptedWithPmk' must match the SecurityProfile type of the source 'EncryptedVMGuestStateOnlyWithPmk'."
```

Because the source VM will always be using guest state only encryption type, I think we should set this based on the value supplied in the confidential_vm_image_encryption_type field, and only assume Guest State if its unset, or maybe better yet require the user to explicitly select an encryption type, what do you think?

Reply:
@JenGoldstrich Thanks for bringing this up, at my first go at this I assumed that `VMGuestStateOnly` and `DiskWithVMGuestState` were mutually exclusive to either PaaS key encryption or CMK key encryption. After checking these errors out and doing some research, I agree that users should be able to supply this setting. Then this setting will be used for either PaaS or CMK-based encryption. I prefer this option to pulling from the `confidential_vm_image_encryption_type` field since that field is tightly coupled to the SIG image version.

I will get this implemented and tested tomorrow.

Reply:
Hey @julian1059, taking another look now

That's fair, my concern is that now `security_encryption_type` is tightly coupled with confidential_vm_image_encryption type, there are a lot of cases where builds can fail now that we'd have to add extra validation cases for, or document around, but if we infer one of these fields from the other then users don't have to set as many fields to build a confidential VM, and there's less errant states for the template to be in.  It feels like this feature requires a lot of fields to be set, and users have to figure out that EncryptedWithPMK expects `DiskWithVMGuestState`.

One way to simplify this in my mind would be as a SIG Destination option to have a field if the image verison is encrypted with a PMK or CMK, and then infer the VM Guest State from the security_encryption_type, since it must always match the encryption state of the image version.  Maybe a field like `use_platform_key_encryption`, what do you think?

Reply:
Also related but I was able to build both VMGuestStateOnly and DiskWithVMGuestState images via PMKs today with this build, nice work!

Reply:
Actually after thinking about this more we should probably keep these separated and just add more validation around this to ensure that there isn't a mismatch between these fields, especially if Azure ever adds other encryption types which they can always do this keeps us more flexible 🤔 , I'll leave my original message just to keep the thought process but I think instead we could just ensure that there isn't a mismatch of the SIG encryption type and the security_encryption_type in the config validation.

Reply:
It's awesome to hear that you successfully built both VMGuestStateOnly and DiskWithVMGuestState images via PMKs! 👌 

Regarding security_type & security_encryption_type, thank you for outlining your thought process here. I agree with your last note, we should leave these two fields separate. The original goal was to simplify as much as possible, but the simplification seems to require assumptions, so yes keeping these fields separate seems ideal at this point.

Regarding validations, are there additional validations you have found other than the ones I listed below? Also, what happens if the inputs are unknown constants to the SDK (for future proofing scenarios), for example, say, `EncryptedVMGuestStateOnlyWithCmk`?
```go
// DiskWithVMGuestState needs to match EncryptedWithPMK
// DiskWithVMGuestState needs to match EncryptedWithCMK
// VMGuestStateOnly needs to match EncryptedVMGuestStateOnlyWithPmk
```

Reply:
Your suggested validations sound good to me, I would add validations to check that the three new fields (Security type, security encryption type, and Confidential VM image encryption type) are only the expected values we support from the SDK (i.e. only. EncryptedWithPMK, EncryptedWithCMK, and EncryptedVMGuestStateOnlyWithPmk are allowed for confidential vm image encryption type)

One more validation that might be worth it that I'm less confident in and would like to hear your opinion is validation to only allow `confidential_vm_image_encryption_type` to be set when security type is confidential, since this field is ignored otherwise.

In terms of future proofing if we have a clean error in the validator that says for example for security_encryption_type "Only DiskWithVMGuestState or VMGuestStateOnly is supported", it'll make it easy in the future to quickly identify why this validation was added and where in the code we'd need to change to support a new field, I wouldn't worry about future proofing unknown constants past that, if we combined the variables I think this would be less future proof.